app/src/main/java/tech/relaycorp/gateway/data/preference/PublicGatewayPreferences.kt

```diff
@@ -23,30 +24,44 @@ class PublicGatewayPreferences
         preferences.get().getString("address", DEFAULT_ADDRESS)
     }
 
-    fun getAddress() = { address }.toFlow()
+    suspend fun getAddress() = observeAddress().first()
+    private fun observeAddress() = { address }.toFlow()
     suspend fun setAddress(value: String) = address.setAndCommit(value)
 
     // Certificate
 
     private val certificate by lazy {
-        preferences.get().getString("certificate")
+        preferences.get().getString("public_gateway_certificate")
     }
 
-    fun getCertificate() = { certificate }.toFlow()
+    suspend fun getCertificate() = observeCertificate().first()
+
+    private fun observeCertificate() = { certificate }.toFlow()
         .map {
             val certificateBytes = if (it.isEmpty()) {
-                resources.openRawResource(R.raw.public_gateway_cert).readBytesAndClose()
+                readRawFile.read(R.raw.public_gateway_cert)
             } else {
                 it.toByteArray(Charset.defaultCharset())
             }
             Certificate.deserialize(certificateBytes)
         }
 
     suspend fun setCertificate(value: Certificate) =
-        address.setAndCommit(value.serialize().toString(Charset.defaultCharset()))
+        certificate.setAndCommit(value.serialize().toString(Charset.defaultCharset()))
+
+    // Registration State
+
+    private val registrationState by lazy {
+        preferences.get().getEnum("registration_state", RegistrationState.ToDo)
+    }
+
+    suspend fun getRegistrationState() = observeRegistrationState().first()
+    private fun observeRegistrationState() = { registrationState }.toFlow()
+    suspend fun setRegistrationState(value: RegistrationState) =
+        registrationState.setAndCommit(value)
 
     companion object {
         @VisibleForTesting
-        internal const val DEFAULT_ADDRESS = "https://cogrpc-test.relaycorp.tech"
+        internal const val DEFAULT_ADDRESS = "https://poweb-test.relaycorp.tech"
```

Comment:
We're in a weird transition right now because I haven't finished the deployment of the new environment, so we actually need both addresses for the time being.

`cogrpc-test.relaycorp.tech` is CogRPC server, so it's the recipient of the cargo. `poweb-test.relaycorp.tech` is the PoWeb server.

As soon as I've finished what I'm doing, we'll be able to have a single address: `test.relaycorp.tech`. Behind the scenes, we're going to be using DNS SRV records to map:

- The CogRPC address for `test.relaycorp.tech` to something like `cogrpc-test.relaycorp.tech`.
- The PoWeb address for `test.relaycorp.tech` to something like `poweb-test.relaycorp.tech`.

I'll take care of that migration -- I basically need to create the new SRV records, get the CogRPC/PoWeb libs to resolve them and update this class to ensure we're only using one address in the preferences.

For now, please introduce a new constant for the PoWeb address.

(The public gateway certificate is the same across both servers; it's basically one server with two different interfaces)

Reply:
I distinguished instead the CogRPC url.